File: app/src/main/java/io/github/xalrk/nudge/ui/Die3D.kt
package io.github.xalrk.nudge.ui

import androidx.compose.foundation.Canvas
import androidx.compose.material3.LocalContentColor
import androidx.compose.material3.MaterialTheme
import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.Path
import androidx.compose.ui.graphics.drawscope.DrawScope
import androidx.compose.ui.graphics.drawscope.Stroke
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.semantics
import kotlin.math.cos
import kotlin.math.sin

/**
 * A wireframe die rendered as a real cube. [rotX]/[rotY] are in degrees; at the angles from
 * [anglesFor] a single face is flat toward the viewer, which is how a roll ends.
 */
@Composable
fun Die3D(rotX: Float, rotY: Float, modifier: Modifier = Modifier, description: String = "Die") {
    val line = LocalContentColor.current
    val fill = MaterialTheme.colorScheme.surface
    Canvas(modifier.semantics { contentDescription = description }) { drawDie(rotX, rotY, line, fill) }
}

/** Rotation (x, y) in degrees that presents [face] (1..6) squarely to the viewer. */
fun anglesFor(face: Int): Pair<Float, Float> = when (face.coerceIn(1, 6)) {
    1 -> 0f to 0f        // +Z front
    2 -> 0f to -90f      // +X right
    3 -> 90f to 0f       // +Y top (y up in model space)
    4 -> -90f to 0f      // -Y bottom
    5 -> 0f to 90f       // -X left
    else -> 0f to 180f   // -Z back
}

private class V3(val x: Float, val y: Float, val z: Float)

private fun rotate(v: V3, ax: Float, ay: Float): V3 {
    // Rotate about X, then about Y.
    val cx = cos(ax); val sx = sin(ax)
    val y1 = v.y * cx - v.z * sx
    val z1 = v.y * sx + v.z * cx
    val cy = cos(ay); val sy = sin(ay)
    val x2 = v.x * cy + z1 * sy
    val z2 = -v.x * sy + z1 * cy
    return V3(x2, y1, z2)
}

/** Faces as (normal, u axis, v axis, pip count). Opposite faces sum to seven. */
private val FACES: List<Triple<V3, Pair<V3, V3>, Int>> = listOf(
    Triple(V3(0f, 0f, 1f), V3(1f, 0f, 0f) to V3(0f, 1f, 0f), 1),
    Triple(V3(1f, 0f, 0f), V3(0f, 0f, -1f) to V3(0f, 1f, 0f), 2),
    Triple(V3(0f, 1f, 0f), V3(1f, 0f, 0f) to V3(0f, 0f, -1f), 3),
    Triple(V3(0f, -1f, 0f), V3(1f, 0f, 0f) to V3(0f, 0f, 1f), 4),
    Triple(V3(-1f, 0f, 0f), V3(0f, 0f, 1f) to V3(0f, 1f, 0f), 5),
    Triple(V3(0f, 0f, -1f), V3(-1f, 0f, 0f) to V3(0f, 1f, 0f), 6),
)

private fun pips(n: Int): List<Pair<Float, Float>> {
    val o = 0.5f
    return when (n) {
        1 -> listOf(0f to 0f)
        2 -> listOf(-o to -o, o to o)
        3 -> listOf(-o to -o, 0f to 0f, o to o)
        4 -> listOf(-o to -o, o to -o, -o to o, o to o)
        5 -> listOf(-o to -o, o to -o, 0f to 0f, -o to o, o to o)
        else -> listOf(-o to -o, o to -o, -o to 0f, o to 0f, -o to o, o to o)
    }
}

private fun DrawScope.drawDie(rotXDeg: Float, rotYDeg: Float, line: Color, fill: Color) {
    val ax = Math.toRadians(rotXDeg.toDouble()).toFloat()
    val ay = Math.toRadians(rotYDeg.toDouble()).toFloat()
    val half = size.minDimension * 0.30f
    val camera = size.minDimension * 2.2f
    val center = Offset(size.width / 2, size.height / 2)
    fun project(v: V3): Offset {
        val r = rotate(V3(v.x * half, v.y * half, v.z * half), ax, ay)
        val s = camera / (camera - r.z)
        return Offset(center.x + r.x * s, center.y - r.y * s)
    }
    val stroke = size.minDimension * 0.075f
    val visible = FACES.map { it to rotate(it.first, ax, ay).z }.filter { it.second > 0.02f }.sortedBy { it.second }
    for ((face, _) in visible) {
        val (n, axes, count) = face
        val (u, v) = axes
        fun corner(su: Float, sv: Float) = project(V3(n.x + u.x * su + v.x * sv, n.y + u.y * su + v.y * sv, n.z + u.z * su + v.z * sv))
        val path = Path().apply {
            moveTo(corner(-1f, -1f).x, corner(-1f, -1f).y)
            lineTo(corner(1f, -1f).x, corner(1f, -1f).y)
            lineTo(corner(1f, 1f).x, corner(1f, 1f).y)
            lineTo(corner(-1f, 1f).x, corner(-1f, 1f).y)
            close()
        }
        drawPath(path, fill)
        drawPath(path, line, style = Stroke(stroke, join = androidx.compose.ui.graphics.StrokeJoin.Round))
        val facing = rotate(n, ax, ay).z.coerceIn(0f, 1f)
        for ((pu, pv) in pips(count)) {
            val p = project(V3(n.x + u.x * pu + v.x * pv, n.y + u.y * pu + v.y * pv, n.z + u.z * pu + v.z * pv))
            drawCircle(line, radius = size.minDimension * 0.07f * (0.6f + 0.4f * facing), center = p)
        }
    }
}

File: ios/Nudge/Sources/UI/Die3D.swift
import SwiftUI

/// A wireframe die rendered as a real cube. rotX/rotY are in degrees; at the angles from
/// `anglesFor` a single face is flat toward the viewer, which is how a roll ends.
struct Die3D: View {
    @Environment(\.theme) private var theme
    let rotX: Double
    let rotY: Double
    var line: Color? = nil

    var body: some View {
        Canvas { ctx, size in
            drawDie(&ctx, size, rotX, rotY, line ?? theme.onSurface, theme.background)
        }
        .accessibilityLabel("Re-roll times")
    }
}

/// Rotation (x, y) in degrees that presents `face` (1..6) squarely to the viewer.
func anglesFor(_ face: Int) -> (Double, Double) {
    switch max(1, min(6, face)) {
    case 1: return (0, 0)
    case 2: return (0, -90)
    case 3: return (90, 0)
    case 4: return (-90, 0)
    case 5: return (0, 90)
    default: return (0, 180)
    }
}

private struct V3 { var x: Double; var y: Double; var z: Double }

private func rotate(_ v: V3, _ ax: Double, _ ay: Double) -> V3 {
    let cx = cos(ax), sx = sin(ax)
    let y1 = v.y * cx - v.z * sx
    let z1 = v.y * sx + v.z * cx
    let cy = cos(ay), sy = sin(ay)
    let x2 = v.x * cy + z1 * sy
    let z2 = -v.x * sy + z1 * cy
    return V3(x: x2, y: y1, z: z2)
}

/// Faces as (normal, u axis, v axis, pip count). Opposite faces sum to seven.
private let faces: [(V3, V3, V3, Int)] = [
    (V3(x: 0, y: 0, z: 1), V3(x: 1, y: 0, z: 0), V3(x: 0, y: 1, z: 0), 1),
    (V3(x: 1, y: 0, z: 0), V3(x: 0, y: 0, z: -1), V3(x: 0, y: 1, z: 0), 2),
    (V3(x: 0, y: 1, z: 0), V3(x: 1, y: 0, z: 0), V3(x: 0, y: 0, z: -1), 3),
    (V3(x: 0, y: -1, z: 0), V3(x: 1, y: 0, z: 0), V3(x: 0, y: 0, z: 1), 4),
    (V3(x: -1, y: 0, z: 0), V3(x: 0, y: 0, z: 1), V3(x: 0, y: 1, z: 0), 5),
    (V3(x: 0, y: 0, z: -1), V3(x: -1, y: 0, z: 0), V3(x: 0, y: 1, z: 0), 6),
]

private func pips(_ n: Int) -> [(Double, Double)] {
    let o = 0.5
    switch n {
    case 1: return [(0, 0)]
    case 2: return [(-o, -o), (o, o)]
    case 3: return [(-o, -o), (0, 0), (o, o)]
    case 4: return [(-o, -o), (o, -o), (-o, o), (o, o)]
    case 5: return [(-o, -o), (o, -o), (0, 0), (-o, o), (o, o)]
    default: return [(-o, -o), (o, -o), (-o, 0), (o, 0), (-o, o), (o, o)]
    }
}

private func drawDie(_ ctx: inout GraphicsContext, _ size: CGSize, _ rotXDeg: Double, _ rotYDeg: Double, _ line: Color, _ fill: Color) {
    let ax = rotXDeg * .pi / 180
    let ay = rotYDeg * .pi / 180
    let minDim = min(size.width, size.height)
    let half = minDim * 0.30
    let camera = minDim * 2.2
    let center = CGPoint(x: size.width / 2, y: size.height / 2)
    func project(_ v: V3) -> CGPoint {
        let r = rotate(V3(x: v.x * half, y: v.y * half, z: v.z * half), ax, ay)
        let s = camera / (camera - r.z)
        return CGPoint(x: center.x + r.x * s, y: center.y - r.y * s)
    }
    let stroke = minDim * 0.075
    let visible = faces.map { ($0, rotate($0.0, ax, ay).z) }.filter { $0.1 > 0.02 }.sorted { $0.1 < $1.1 }
    for (face, _) in visible {
        let (n, u, v, count) = face
        func corner(_ su: Double, _ sv: Double) -> CGPoint {
            project(V3(x: n.x + u.x * su + v.x * sv, y: n.y + u.y * su + v.y * sv, z: n.z + u.z * su + v.z * sv))
        }
        var path = Path()
        path.move(to: corner(-1, -1)); path.addLine(to: corner(1, -1)); path.addLine(to: corner(1, 1)); path.addLine(to: corner(-1, 1)); path.closeSubpath()
        ctx.fill(path, with: .color(fill))
        ctx.stroke(path, with: .color(line), style: StrokeStyle(lineWidth: stroke, lineJoin: .round))
        let facing = min(max(rotate(n, ax, ay).z, 0), 1)
        for (pu, pv) in pips(count) {
            let p = project(V3(x: n.x + u.x * pu + v.x * pv, y: n.y + u.y * pu + v.y * pv, z: n.z + u.z * pu + v.z * pv))
            let r = minDim * 0.07 * (0.6 + 0.4 * facing)
            ctx.fill(Path(ellipseIn: CGRect(x: p.x - r, y: p.y - r, width: 2 * r, height: 2 * r)), with: .color(line))
        }
    }
}
